Add an OpenMP dependency.

QuLogic opened the pull request:
Unfortunately, it seems like every compiler (mostly Fortran ones) use a different flag for OpenMP. This works similarly to the thread dependency which stores the various inconsistent flags in each compiler.

The problem though is that, like the thread dependency, it just assumes that OpenMP is supported. I'm not exactly sure how old the compilers are that don't support it though (at least gcc 4.2+, Intel 12+). Also, what to do about other languages like Vala or D?

jpakkane commented:
How about adding a method in the base class that throws an exception saying openmp is not supported on this language and adding an override to the Vala class that returns `[]`?

QuLogic commented:
But what would be calling that method?

jpakkane commented:
If someone accidentally puts an openmp dependency on a Jar target, for example.

QuLogic commented:
Right, but I mean, this would be checked at build time or configure time? It'd be preferable to be the latter, but I'm not sure where to put that.

jpakkane commented:
It should fail when generating backend files, right? So a basic failing test.

Why is the test code checking for the value 2? Is is the default for openmp? The test suite needs to run on platforms with arbitrary number of CPUs and one would imagine that on a 48 core machine it would return 48.

This also needs the docs update and mention in the release notes because this is cool new stuff.

Timmmm commented:
Apple's LLVM *still* doesn't support OpenMP so you definitely can't assume it.

QuLogic commented:
@Timmmm ah, I didn't realize the framework tests aren't run on macOS. I've changed it to run everywhere, but if as you say they don't support it, then it'll fail as I haven't added any checks yet.

@jpakkane It checks for 2 because that's what I set in the env.

jpakkane commented:
This should probably test for `has_header` with the openmp dependency and skip the test if it is not found.

QuLogic commented:
You know, you can get away with a lot of OpenMP stuff without the header, so I almost forgot about its existence even though it's in the test. This should make it a bit easier to check compiler support at least.

QuLogic commented:
Clang is weird. It supports the directives with `-fopenmp` but provides no runtime library.  You can specify the library with `-fopenmp=something`, but they aren't located in any standard location. They seem to require you to manually specify the `CFLAGS` and/or `LDFLAGS` in order to find it. Theoretically, that means you could build with OpenMP on macOS if you built the library and put the location in the search flags.

amitdo commented:
http://releases.llvm.org/3.8.0/tools/clang/docs/ReleaseNotes.html#openmp-support-in-clang
http://releases.llvm.org/3.7.0/tools/clang/docs/ReleaseNotes.html#openmp-support
http://openmp.llvm.org/

amitdo commented:
@QuLogic, can you update this so it can be merged?

QuLogic commented:
I've updated this and added some explicit checks now. It works locally with gcc 7.3.1 and clang 5.0.1.

@jpakkane can you add `libiomp-dev` to the CI image? It adds the OpenMP header for clang.

I do not know what's up with macOS builds; it says clang 8, but the flag doesn't work. I guess Apple still haven't merged this stuff, even though it's in upstream clang? I suppose I'll have to skip it.

For MSVC, it looks like I missed a flag or two that I will look into later.

amitdo commented:
I do not know what's up with macOS builds; it says clang 8, but the flag doesn't work. I guess Apple still haven't merged this stuff, even though it's in upstream clang?

The OpenMP runtime is not shipped with Xcode's Clang. So you need to get and install llvm/gcc OpenMP runtime first.

https://github.com/tesseract-ocr/tesseract/wiki/Compiling#prepare-support-for-openmp-optional

QuLogic commented:
I'm not sure if we want to do that; for now I've added a skip, but it can be re-enabled if wanted. All that's left is to update the docker image with clang's omp library.

jpakkane commented:
Docker image updated.

QuLogic commented:
Great, looks to all pass now.

QuLogic commented:
Rebased due to conflicts.
